openff/qcsubmit/utils/utils.py

```diff
@@ -1,16 +1,256 @@
+        """
+        if missing_ok:
+            logger.warning("missing_ok provided but unused by CachedPortalClient")
+        if unpack := not isinstance(record_ids, Sequence):
+            record_ids = [record_ids]
+        res = get_records_with_cache(
+            client=self,
+            record_cache=self.record_cache,
+            record_type=SinglepointRecord,
+            record_ids=record_ids,
+            include=include,
+            force_fetch=False,
+        )
+        if unpack:
+            return res[0]
+        else:
+            return res
+
+    def get_torsiondrives(
+        self,
+        record_ids: Union[int, Sequence[int]],
+        missing_ok: bool = False,
+        *,
+        include: Optional[Iterable[str]] = None,
+    ) -> Union[Optional[TorsiondriveRecord], List[Optional[TorsiondriveRecord]]]:
+        """
+        Obtain torsiondrive records with the specified IDs.
+
+        Records will be returned in the same order as the record ids.
+
+        Parameters
+        ----------
+        record_ids
+            Single ID or sequence/list of records to obtain
+        missing_ok
+            Unlike a ``PortalClient``, this argument is ignored. If set to
+            True, a warning will be printed. Any missing records will cause a
+            ``RuntimeError`` to be raised.
+        include
+            Additional fields to include in the returned record
+
+        Returns
+        -------
+        :
+            If a single ID was specified, returns just that record. Otherwise,
+            returns a list of records.
+        """
+        if missing_ok:
+            logger.warning("missing_ok provided but unused by CachedPortalClient")
+        if unpack := not isinstance(record_ids, Sequence):
+            record_ids = [record_ids]
+        res = get_records_with_cache(
+            client=self,
+            record_cache=self.record_cache,
+            record_type=TorsiondriveRecord,
+            record_ids=record_ids,
+            include=include,
+            force_fetch=False,
+        )
+        if unpack:
+            return res[0]
+        else:
+            return res
+
+    @contextmanager
+    def _no_session(self):
+        """This is a supplemental context manager to the ``no_internet``
+        manager in _tests/utils/test_manager.py. ``PortalClient`` creates a
+        ``requests.Session`` on initialization that can be reused without
+        accessing ``socket.socket`` again. Combining ``no_internet`` and
+        ``client._no_session`` should completely ensure that the local cache is
+        used rather than re-fetching data from QCArchive.
+        """
+        tmp = self._req_session
+        self._req_session = None
+        try:
+            yield
+        finally:
+            self._req_session = tmp
 
 
 def _default_portal_client(client_address) -> PortalClient:
-    return PortalClient(client_address)
+    return CachedPortalClient(client_address, cache_dir=_DEFAULT_CACHE_DIR)
```

Comment:
Good call on this. This code fails with a `database is locked` error:

```python
from multiprocessing import Pool

from qcportal import PortalClient

from openff.qcsubmit.results import TorsionDriveResultCollection
from openff.qcsubmit.utils import CachedPortalClient, portal_client_manager

ds = TorsionDriveResultCollection.parse_file(
    "/home/brent/omsf/projects/valence-fitting/02_curate-data/datasets/supp-td.json"
)
datasets = [ds, ds, ds, ds]
client = CachedPortalClient("https://api.qcarchive.molssi.org:443", cache_dir=".")
with portal_client_manager(lambda _: client), Pool(4) as pool:
    for res in pool.imap_unordered(
        TorsionDriveResultCollection.to_records, datasets, 1
    ):
        print(len(res))
```

It also fails without the `portal_client_manager` and `CachedPortalClient` as the default client, so it's not safe to share the same client across threads or to construct multiple clients accessing the same database.

It actually doesn't quite work most of the time, even with a regular `PortalClient` with or without a cache dir, but some of the processes have `read` timeout errors rather than the more dramatic database errors from the cached version. It's probably best in general not to share a `PortalClient` across threads.

Using `with portal_client_manager(PortalClient), Pool(4) as pool:` (removing the lambda that shares the same client), as your suggestion will restore the default to, has worked successfully on every run I've tried so far, so I think that's definitely the right default.

Do you think it's worth mentioning something about this in the documentation for `portal_client_manager`? The docs currently show an example of a function that returns a new `PortalClient` each time (which should be safe) rather than a `lambda _: client`. I'm not really sure how to warn against this without showing the lambda code in the docs, though.

Reply:
Thanks for the update and for checking on multithreading. I've read this a few times and think I kinda understand, and my preference is to do the simplest safe thing. So I agree with reverting the default behavior. And updating the docs is optional, you might just copy your comment "It's not safe to share the same client across threads or to construct multiple clients accessing the same database" into the portal_client_manager docstring and call it a day.